x/mongo/driver/topology/server_test.go

```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+				return context.Background(), nil
+			},
+			expectErr:         false,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "Parent context deadline exceeded error during dialing should not clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
+			},
+			dialer: func(Dialer) Dialer {
+				var d net.Dialer
+				return DialerFunc(func(ctx context.Context, network, addr string) (net.Conn, error) {
+					// Sleep for at least 150ms and expect the context passed to server.Connection()
+					// to time out during the sleep. Expect the error returned by DialContext() to
+					// be treated as a timeout caused by an operation-scoped deadline.
+					// E.g. FindOne(context.WithTimeout(...))
+					time.Sleep(150 * time.Millisecond)
+					return d.DialContext(ctx, network, addr)
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "Child context deadline exceeded error during dialing should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.Background(), nil
+			},
+			dialer: func(Dialer) Dialer {
+				var d net.Dialer
+				return DialerFunc(func(ctx context.Context, network, addr string) (net.Conn, error) {
+					// Wrap the context in a context with an already-exceeded deadline. Expect the
+					// error returned by DialContext() to be treated as a timeout caused by reaching
+					// connectTimeoutMS.
+					ctx, cancel := context.WithDeadline(ctx, time.Now().Add(-1*time.Hour))
+					defer cancel()
+					return d.DialContext(ctx, network, addr)
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+		{
+			desc: "Parent context deadline exceeded error during handshake should not clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
+			},
+			handshaker: func(Handshaker) Handshaker {
+				h := auth.Handshaker(nil, &auth.HandshakeOptions{})
+				return &testHandshaker{
+					getHandshakeInformation: func(ctx context.Context, addr address.Address, c driver.Connection) (driver.HandshakeInformation, error) {
+						// Sleep for at least 150ms and expect the context passed to
+						// server.Connection() to time out during the sleep. Expect the error
+						// returned by GetHandshakeInformation() to be treated as a timeout caused
+						// by an operation-scoped deadline.
+						// E.g. FindOne(context.WithTimeout(...))
+						time.Sleep(150 * time.Millisecond)
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "Child context deadline exceeded error during handshake should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
```

Comment:
Minor: Since this is only testing that the context within the handshake times out, should this be `context.Background()` for consistency with the "Child context" test above? 
```suggestion
				return context.Background(), nil
```

Reply:
Using a `context.WithTimeout` and a `context.Background` are actually two different test cases where we expect the same behavior (when the error was caused by a child context). I decided to use the `context.WithTimeout` to cover the case that was handled incorrectly by the previous code (i.e. a context with deadline was used to call `Connection()` but was not the source of the timeout error).

I'll add a comment to clarify the intent of the `context.WithTimeout`.